### PR TITLE
Translate SocketError.MessageSize to IPStatus.PacketTooBig in Ping.SendPing[Async]

### LINKED ISSUE #55961: [Regression] Ping.SendPingAsync API throws SocketException: Message too long exception on .NET 6 Preview 6
### Description

PowerShell uses the API `System.Net.NetworkInformation.Ping.SendPingAsync` for the Test-Connection cmdlet. 

Only on macOS and on .NET 6 Preview 6, we get an exception when trying to find the supported MTUSize.

The exception is a System.Net.NetworkInformation.PingException with an inner exception of System.Net.Sockets.SocketException with a message "Message too long".

### Configuration

.NET 6 Preview 6 - osx-x64.

### Regression?

This is a regression from .NET 6 preview 4. We did not test on Preview 5.

### Other information

Exception :
Type : System.Net.NetworkInformation.PingException
Message : Testing connection to computer '127.0.0.1' failed: An exception occurred during a Ping request.
InnerException :
Type : System.Net.Sockets.SocketException
Message : Message too long
SocketErrorCode : MessageSize
ErrorCode : 40
NativeErrorCode : 40
TargetSite :
Name : Throw
DeclaringType : System.Runtime.ExceptionServices.ExceptionDispatchInfo
MemberType : Method
Module : System.Private.CoreLib.dll
Source : System.Private.CoreLib
HResult : -2147467259
StackTrace :

```
at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.CreateException(SocketError error, Boolean forAsyncThrow) in System.Net.Sockets.dll:token 0x60002ba+0xa
at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.SendToAsync(Socket socket, CancellationToken cancellationToken) in System.Net.Sockets.dll:token 0x60002ac+0x32
at System.Net.Sockets.Socket.SendToAsync(ReadOnlyMemory`1 buffer, SocketFlags socketFlags, EndPoint remoteEP, CancellationToken cancellationToken) in System.Net.Sockets.dll:token 0x6000286+0x51
at System.Net.NetworkInformation.Ping.SendIcmpEchoRequestOverRawSocketAsync(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options) in System.Net.Ping.dll:token 0x600004b+0xa2
at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in System.Private.CoreLib.dll:token 0x6004877+0x28
at System.Net.NetworkInformation.Ping.SendIcmpEchoRequestOverRawSocketAsync(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options) in System.Net.Ping.dll:token 0x600004b+0x3d
at System.Net.NetworkInformation.Ping.SendPingAsyncCore(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options) in System.Net.Ping.dll:token 0x6000053+0xe
at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in System.Private.CoreLib.dll:token 0x6004877+0x28
at System.Net.NetworkInformation.Ping.SendPingAsyncCore(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options) in System.Net.Ping.dll:token 0x6000053+0x3d
at System.Net.NetworkInformation.Ping.SendPingAsyncInternal(IPAddress address, Int32 timeout, Byte[] buffer, PingOptions options) in System.Net.Ping.dll:token 0x6000040+0x24
at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in System.Private.CoreLib.dll:token 0x6004877+0x28
at System.Net.NetworkInformation.Ping.SendPingAsyncInternal(IPAddress address, Int32 timeout, Byte[] buffer, PingOptions options) in System.Net.Ping.dll:token 0x6000040+0x3d
at System.Net.NetworkInformation.Ping.SendPingAsync(IPAddress address, Int32 timeout, Byte[] buffer, PingOptions options) in System.Net.Ping.dll:token 0x600003f+0x0
at Microsoft.PowerShell.Commands.TestConnectionCommand.SendCancellablePing(IPAddress targetAddress, Int32 timeout, Byte[] buffer, PingOptions pingOptions, Stopwatch timer) in
/Users/travisplunk/git/powershell/src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:line 863
at Microsoft.PowerShell.Commands.TestConnectionCommand.ProcessMTUSize(String targetNameOrAddress) in /Users/travisplunk/git/powershell/src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:line 515
at Microsoft.PowerShell.Commands.TestConnectionCommand.ProcessRecord() in /Users/travisplunk/git/powershell/src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs:line 260
at System.Management.Automation.Cmdlet.DoProcessRecord() in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/cmdlet.cs:line 173
at System.Management.Automation.CommandProcessor.ProcessRecord() in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/CommandProcessor.cs:line 387
at System.Management.Automation.CommandProcessorBase.DoExecute() in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/CommandProcessorBase.cs:line 568
at System.Management.Automation.Internal.PipelineProcessor.Inject(Object input, Boolean enumerate) in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/pipeline.cs:line 1165
at System.Management.Automation.Internal.PipelineProcessor.SynchronousExecuteEnumerate(Object input) in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/pipeline.cs:line 495
at System.Management.Automation.PipelineOps.InvokePipeline(Object input, Boolean ignoreInput, CommandParameterInternal[][] pipeElements, CommandBaseAst[] pipeElementAsts, CommandRedirection[][] commandRedirections, FunctionContext
funcContext) in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs:line 498
at System.Management.Automation.Interpreter.ActionCallInstruction`6.Run(InterpretedFrame frame) in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/interpreter/CallInstruction.Generated.cs:line 608
at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame) in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/interpreter/ControlFlowInstructions.cs:line 358
at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame) in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/interpreter/ControlFlowInstructions.cs:line 358
at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame) in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/interpreter/ControlFlowInstructions.cs:line 358
at System.Management.Automation.Interpreter.Interpreter.Run(InterpretedFrame frame) in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/interpreter/Interpreter.cs:line 105
at System.Management.Automation.Interpreter.LightLambda.RunVoid1[T0](T0 arg0) in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/interpreter/LightLambda.Generated.cs:line 81
at System.Management.Automation.DlrScriptCommandProcessor.RunClause(Action`1 clause, Object dollarUnderbar, Object inputToProcess) in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/ScriptCommandProcessor.cs:line
547
at System.Management.Automation.DlrScriptCommandProcessor.Complete() in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/ScriptCommandProcessor.cs:line 425
at System.Management.Automation.CommandProcessorBase.DoComplete() in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/CommandProcessorBase.cs:line 642
at System.Management.Automation.Internal.PipelineProcessor.DoCompleteCore(CommandProcessorBase commandRequestingUpstreamCommandsToStop) in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/pipeline.cs:line 603
at System.Management.Automation.Internal.PipelineProcessor.SynchronousExecuteEnumerate(Object input) in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/pipeline.cs:line 514
at System.Management.Automation.Runspaces.LocalPipeline.InvokeHelper() in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs:line 400
at System.Management.Automation.Runspaces.LocalPipeline.InvokeThreadProc() in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs:line 591
at System.Management.Automation.Runspaces.PipelineThread.WorkerProc() in /Users/travisplunk/git/powershell/src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs:line 1220
at System.Threading.Thread.StartHelper.Callback(Object state) in System.Private.CoreLib.dll:token 0x60027aa+0x0
at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in System.Private.CoreLib.dll:token 0x6002968+0x40
at System.Threading.Thread.StartCallback() in System.Private.CoreLib.dll:token 0x600272e+0xe
--- End of stack trace from previous location ---
at System.Net.NetworkInformation.Ping.SendIcmpEchoRequestOverRawSocketAsync(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options) in System.Net.Ping.dll:token 0x600004b+0x11f
at System.Net.NetworkInformation.Ping.SendPingAsyncCore(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options) in System.Net.Ping.dll:token 0x6000053+0x89
at System.Net.NetworkInformation.Ping.SendPingAsyncInternal(IPAddress address, Int32 timeout, Byte[] buffer, PingOptions options) in System.Net.Ping.dll:token 0x6000040+0x9c
```

### PR DESCRIPTION
**Pull request by filipnavara**
Fixes #55961

### CODE REVIEW

File: src/libraries/System.Net.Ping/tests/FunctionalTests/PingTest.cs [hunk 1]
```diff
@@ -647,6 +647,20 @@ public async Task SendPingAsyncWithHostAndTimeoutAndBufferAndPingOptions_Unix()
                 });
         }
 
+        [Fact]
+        [PlatformSpecific(TestPlatforms.OSX)]
```
Comment (stephentoub) [inline src/libraries/System.Net.Ping/tests/FunctionalTests/PingTest.cs:651]: There's no equivalent behavior on other platforms?

**Comment by stephentoub** [comment]
> Fixes #55961

It'd be good to follow-up after this to see whether there are other error codes we should be translating as well.  Needn't be you, of course, just suggesting we have an issue tracking it.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.RawSocket.cs [hunk 2]
```diff
@@ -239,6 +243,10 @@ private async Task<PingReply> SendIcmpEchoRequestOverRawSocketAsync(IPAddress ad
                 catch (SocketException ex) when (ex.SocketErrorCode == SocketError.TimedOut)
                 {
                 }
+                catch (SocketException ex) when (ex.SocketErrorCode == SocketError.MessageSize)
+                {
+                    return new PingReply(new IPAddress(0), null, IPStatus.PacketTooBig, 0, Array.Empty<byte>());
```
Comment (stephentoub) [inline src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.RawSocket.cs:248]: Maybe we should modify the CreateTimedOutPingReply() helper to instead be CreatePingReply(IPStatus status) ?

File: src/libraries/System.Net.Ping/tests/FunctionalTests/PingTest.cs (hunk 1 above)
Comment (filipnavara) [reply src/libraries/System.Net.Ping/tests/FunctionalTests/PingTest.cs:651]: There is but it's somewhat inconsistent and may report timeout instead. I drafted this only to see what happens on the CI.

File: src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.RawSocket.cs (hunk 2 above)
Comment (filipnavara) [reply src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.RawSocket.cs:248]: Yep, I was planning to do that before moving this out of draft.

File: src/libraries/System.Net.Ping/tests/FunctionalTests/PingTest.cs (hunk 1 above)
Comment (filipnavara) [reply src/libraries/System.Net.Ping/tests/FunctionalTests/PingTest.cs:651]: Basically the size limit can differ between platforms and the error reported can be `TimedOut` (on Unix, non-privileged run) or `PacketTooBig` (macOS, iOS, Windows, privileged Unix). I am not sure how much effort will it take to test all the combinations and cover them by the tests so I started small.

**Review by stephentoub — APPROVED** [review]
Thanks!

**Comment by karelz** [comment]
Thanks a lot @filipnavara !

**Pull request merged**